### PR TITLE
Batch inscribe on same sat

### PR DESCRIPTION
**Pull request by raphjaph**
Allow to reinscribe a whole batch onto the same sat. All inscriptions will land on the sat that the first inscription in the batch will land on.
 #2593

### CODE REVIEW

**Comment by elocremarc** [comment]
Will this create a new cursed type? Do we need to consider this in the jubilee?

**Comment by raphjaph** [comment]
Shouldn't require a new curse. This isn't a new protocol feature just a new wallet feature.

File: src/subcommand/wallet/inscribe.rs [hunk 1]
```diff
@@ -172,6 +175,12 @@ impl Inscribe {
         )?;
 
         mode = batchfile.mode;
+
+        reinscribe = mode == Mode::Reinscribe;
+
+        if reinscribe && batchfile.parent.is_some() {
+          return Err(anyhow!("reinscribe mode cannot be used with parent"));
+        }
```
Comment (casey) [inline src/subcommand/wallet/inscribe.rs:183]:
```suggestion
```

File: src/subcommand/wallet/inscribe/batch.rs [hunk 2]
```diff
@@ -520,6 +528,8 @@ impl Batch {
 
 #[derive(PartialEq, Debug, Copy, Clone, Serialize, Deserialize, Default)]
 pub(crate) enum Mode {
+  #[serde(rename = "reinscribe")]
+  Reinscribe,
```
Comment (casey) [inline src/subcommand/wallet/inscribe/batch.rs:532]: Maybe we should call this `SharedSat` or `SameSat`?

**Review by casey — CHANGES_REQUESTED** [review]

File: batch.yaml [hunk 3]
```diff
@@ -3,6 +3,7 @@
 # there are two modes:
 # - `separate-outputs`: place all inscriptions in separate postage-sized outputs
 # - `shared-output`: place inscriptions in a single output separated by postage
+# - `same-sat`: place inscriptions on the same sat
```
Comment (casey) [inline batch.yaml:6]:
```suggestion
# - `separate-outputs`: inscribe on separate postage-sized outputs
# - `shared-output`: inscribe on a single output separated by postage
# - `same-sat`: inscribe on the same sat
```

File: src/subcommand/wallet/inscribe/batch.rs [hunk 4]
```diff
@@ -615,10 +624,11 @@ impl Batchfile {
         compress,
       )?);
 
-      pointer += postage.to_sat();
+      pointer = pointer.map(|value| value + postage.to_sat());
     }
 
     let destinations = match self.mode {
+      Mode::SameSat => vec![get_change_address(client, chain)?],
       Mode::SharedOutput => vec![get_change_address(client, chain)?],
```
Comment (casey) [inline src/subcommand/wallet/inscribe/batch.rs:632]:
```suggestion
      Mode::SameSat | Mode::SharedOutput => vec![get_change_address(client, chain)?],
```

File: src/subcommand/wallet/inscribe/batch.rs [hunk 5]
```diff
@@ -286,6 +292,7 @@ impl Batch {
       .map(|destination| TxOut {
         script_pubkey: destination.script_pubkey(),
         value: match self.mode {
+          Mode::SameSat => self.postage.to_sat(),
           Mode::SeparateOutputs => self.postage.to_sat(),
           Mode::SharedOutput => total_postage.to_sat(),
```
Comment (casey) [inline src/subcommand/wallet/inscribe/batch.rs:297]:
```suggestion
          Mode::SameSat | Mode::SeparateOutputs => self.postage.to_sat(),
          Mode::SharedOutput => total_postage.to_sat(),
```

File: src/subcommand/wallet/inscribe/batch.rs [hunk 6]
```diff
@@ -198,6 +199,11 @@ impl Batch {
     }
 
     match self.mode {
+      Mode::SameSat => assert_eq!(
+        self.destinations.len(),
+        1,
+        "invariant: reinscribe has only one destination"
```
Comment (casey) [inline src/subcommand/wallet/inscribe/batch.rs:205]:
```suggestion
        "invariant: same sat can only inscribe to one destination"
```

**Review by casey — APPROVED** [review]

**Pull request merged**